File: 防火墙项目/FWclient-存文件/FWclient/interfaces-classes/ConfigWhiteLists.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FWclient
{
    class ConfigWhiteLists:IConfigWhiteLists
    {
        private DeviceForm devform;
        FileOperation of = new FileOperation("C:\\CMP\\Config");
        public ConfigWhiteLists()
        {
            this.devform = new DeviceForm("0.0.0.0",22222);
        }

        public bool ChangeWhiteLists(string dev_IP, string dst_IP, string src_IP,  string dst_port, string src_port, bool log_record, bool add_delete)
        {
            this.devform.setDev_IP(dev_IP);
            WhiteLists lists = new WhiteLists();
            lists.setIPAndPort(dst_IP, src_IP, dst_port, src_port);

            string flag = null;
    
            string whiteList_from_client_to_server0 = "iptables -A FORWARD -p tcp -s " + lists.getsrc_IP() + " -d " + lists.getdst_IP() + " --sport " + lists.getsrc_port()
                + " --dport " + lists.getdst_port() + " -j ACCEPT ";
            // string whiteList_from_client_to_server1 = "iptables -A FORWARD -p tcp -d" + wl.getSrc_IP() + "--sport" + wl.getPort();

            if (add_delete)
            {
                flag = "WHL1";
                string frule = dev_IP+" "+ dst_IP + " "+ src_IP + " "+ dst_port + " "+ src_port;
                of.SaveRules(frule, "whl");
            }

            else
            {
                flag = "WHL0";
                string frule = dev_IP + " " + dst_IP + " " + src_IP + " " + dst_port + " " + src_port;
                of.DeleteRules(frule, "whl");
            }

            string changewl = flag + whiteList_from_client_to_server0;
            //DeviceForm devform = new DeviceForm(dev_IP, 22222);
            SendInfo sendcmd = new SendInfo(devform);
            return sendcmd.SendConfigInfo(changewl);                           

        }
    }
}

File: QT-firewall/firewall-test/configwhitelists.cpp
#include "configwhitelists.h"

using namespace std;


configwhitelists::configwhitelists(const string &devIP, const string &fw_mac, QObject *parent)
    :FwOrProIP(devIP),fwMAC(fw_mac),QCommInfo(parent)
{
    fileOP = new FileOperation();
    this->setBindInfo(getLocal_IP(),30333);
}

configwhitelists::~configwhitelists()
{
    if(fileOP != nullptr)
        delete fileOP;
    fileOP = nullptr;
}

static std::string file_rule("");
static bool fileFlag = true;
void configwhitelists::ChangeWhiteLists(const std::string &dst_ip,
                                        const std::string &src_ip,
                                        const std::string &dst_port,
                                        const std::string &src_port,
                                        bool add_delete)
{
    string pro = "WHL white";
    string rule = "iptables -A FORWARD -p tcp -s "+src_ip+" -d "+dst_ip
            +" --sport "+src_port+" --dport "+dst_port+" -j ACCEPT";
    string reverse_rule = "iptables -A FORWARD -p tcp -s "+dst_ip+" -d "+src_ip
            +" --sport "+dst_port+" --dport "+src_port+" -j ACCEPT";
    string log_rule = "iptables -A FORWARD -p tcp -s "+src_ip+" -d "+dst_ip
            +" --sport "+src_port+" --dport "+dst_port+" -j LOG --log-prefix "+"\"ACCEPT&WHL \"";
    file_rule = pro+" "+fwMAC+" "+dst_ip+" "+src_ip+" "+dst_port+" "+src_port;
    //cout<<file_rule<<endl;
    string flag = "WHL0";
    if(add_delete)
    {
        flag.replace(flag.begin(),flag.end(),"WHL1");
        fileFlag = true;
    }
    else
    {
        fileFlag = false;
    }
    rule = flag + rule + " && " +reverse_rule + " && " + log_rule;
    this->sendCmd(FwOrProIP,22222,CommInfo::iptablesCmdTypeByte,rule);
}

void configwhitelists::handResultData(const bool result, const char *commandTypeByte, const char *commandBytes, const int commandLength, const QString &srcIp, const int srcPort)
{
    if (result)
    {
        if (*commandTypeByte == *CommInfo::iptablesCmdTypeByte)
        {
            std::string data(commandBytes,commandLength);
            std::cout << data << std::endl;
            if(data == "success")
            {
                if(fileFlag)
                    fileOP->addRule(file_rule);
                else
                    fileOP->deleteRule(file_rule);
                emit configResult(true);
            }

            if(data == "fail")
                emit configResult(false);
        }
    }
}
